### PR TITLE
Fixes text processing bugs

### LINKED ISSUE #7502: Text processing bugs
This issue helps track bugs related to recent text processing changes. Feel free to post your findings here.

### PR DESCRIPTION
**Pull request by Gillibald**
## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #7502

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018693-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Visuals/Media/TextFormatting/TextLineImpl.cs [hunk 1]
```diff
@@ -112,9 +112,15 @@ public override TextLine Collapse(params TextCollapsingProperties[] collapsingPr
...
+            
+            if (collapsingProperties.Width < shapedSymbol.GlyphRun.Size.Width)
+            {
+                return new TextLineImpl(new List<ShapedTextCharacters>(0), textRange, _paragraphWidth, _paragraphProperties,
+                    _flowDirection, TextLineBreak, true);
+            }
+            
             var availableWidth = collapsingProperties.Width - shapedSymbol.GlyphRun.Size.Width;
-
+
```
Comment (YohDeadfall) [inline src/Avalonia.Visuals/Media/TextFormatting/TextLineImpl.cs:123]: nit: whitespaces instead of an empty line.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018777-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018788-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018789-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Visuals/Media/TextFormatting/ShapedTextCharacters.cs [hunk 2]
```diff
@@ -156,6 +154,8 @@ internal SplitResult<ShapedTextCharacters> Split(int length)
 
             if (first.Text.Length != length)
             {
+                Debugger.Break();
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/TextFormatting/ShapedTextCharacters.cs:157]: Remove this

File: src/Avalonia.Visuals/Media/TextFormatting/TextLineBreak.cs [hunk 3]
```diff
@@ -7,6 +9,13 @@ public class TextLineBreak
         public TextLineBreak(TextEndOfLine? textEndOfLine = null, FlowDirection flowDirection = FlowDirection.LeftToRight, 
             IReadOnlyList<ShapedTextCharacters>? remainingCharacters = null)
         {
+            #if DEBUG
+            if (remainingCharacters != null && remainingCharacters.Any(x => x == null))
+            {
+                Debugger.Break();
```
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/TextFormatting/TextLineBreak.cs:15]: Remove this

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018807-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018808-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018837-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018855-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by adirh3** [comment]
@Gillibald these changes really messed up spacing - 
Before -
image

After - 
image

**Comment by Gillibald** [comment]
@adirh3 Y this does not look right. My changes produce slightly bigger text bounds. In the worst case `+1px` I might need to change this and only apply rounding in the arrange step and only do that if we have TextTrimming enabled

**Comment by Gillibald** [comment]
Screenshot 2022-02-23 065102

Each item of this ListBox has a Grid and 3 controls inside that Grid. The last child is a Border with a fixed height. I can't see any major issues.

@adirh3 I think you need to provide me some XAML here. You can also fire up the DevTools

**Review by MarchingCube — APPROVED** [review]
Tested this in my project an it fixed a bunch of problems that we had when using version in master.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0018902-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**